giraffe/src/utils/getLatestValues.ts

```diff
@@ -0,0 +1,127 @@
+import {range} from './range'
+import {isString} from './isString'
+import {flatMap} from './flatMap'
+import {Table, NumericColumnData} from '../types'
+
+/*
+  Return a list of the maximum elements in `xs`, where the magnitude of each
+  element is computed using the passed function `d`.
+*/
+const maxesBy = <X>(xs: X[], d: (x: X) => number): X[] => {
+  let maxes = []
+  let maxDist = -Infinity
+
+  for (const x of xs) {
+    const dist = d(x)
+
+    if (dist > maxDist) {
+      maxes = [x]
+      maxDist = dist
+    } else if (dist === maxDist && dist !== -Infinity) {
+      maxes.push(x)
+    }
+  }
+
+  return maxes
+}
+
+const EXCLUDED_COLUMNS = new Set([
+  '_start',
+  '_stop',
+  '_time',
+  'table',
+  'result',
+  '',
+])
+
+/*
+  Determine if the values in a column should be considered in `latestValues`.
+*/
+const isValueCol = (table: Table, colKey: string): boolean => {
+  const columnType = table.getColumnType(colKey)
+  const columnName = table.getColumnName(colKey)
+
+  return (
+    (columnType === 'number' ||
+      columnType === 'time' ||
+      columnType === 'string') &&
+    !EXCLUDED_COLUMNS.has(columnName)
+  )
+}
+
+/*
+  We sort the column keys that we pluck latest values from, so that:
+
+  - Columns named `_value` have precedence
+  - The returned latest values are in a somewhat stable order
+*/
+const sortTableKeys = (keyA: string, keyB: string): number => {
+  if (keyA.includes('_value')) {
+    return -1
+  } else if (keyB.includes('_value')) {
+    return 1
+  } else {
+    return keyA.localeCompare(keyB)
+  }
+}
+
+/*
+  Return a list of the most recent numeric values present in a `Table`.
+
+  This utility searches any numeric column to find values, and uses the `_time`
+  column as their associated timestamp.
+
+  If the table only has one row, then a time column is not needed.
+*/
+export const getLatestValues = (table: Table): number[] => {
+  const valueColsData = table.columnKeys
+    .sort((a, b) => sortTableKeys(a, b))
+    .filter(k => isValueCol(table, k))
+    .map(k => table.getColumn(k)) as number[][]
```

Comment:
in PR timmy mentioned this was pulled over, no need to tweak it now.

Reply:
This is a very good point. Let's drill for performance while the code is fresh in our minds. I'll look into updating this.

Reply:
So, it looks like the best we can do here is to `.filter` first then `.sort`. `.map` will still have to be last and independent because there doesn't seem to be a way to sort a bunch of strings and transform them into something else simultaneously.